Implement iterable for headers

jeenalee opened the pull request:
<!-- Please describe your changes on the following line: -->

These patches make Headers iterable based on #12819. 

The expected wpt results are updated as well. 

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11897 (github issue number if applicable). 

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because web platform tests for Headers already exist. Expected test results are updated with this PR.

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12998)

<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/webidls/IterableIterator.webidl, components/script/dom/mod.rs, components/script/dom/bindings/codegen/Configuration.py, components/script/dom/webidls/TestBindingIterable.webidl, components/script/dom/bindings/iterable.rs, components/script/dom/bindings/codegen/parser/callback-location.patch, components/script/dom/bindings/codegen/CodegenRust.py, components/script/dom/bindings/mod.rs, components/script/dom/bindings/codegen/parser/update.sh, components/script/dom/webidls/TestBindingPairIterable.webidl, components/script/dom/headers.rs, components/script/dom/webidls/Headers.webidl, components/script/dom/testbindingiterable.rs, components/script/dom/bindings/codegen/parser/WebIDL.py, components/script/dom/testbindingpairiterable.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

jeenalee commented:
r? @jdm

jdm commented on components/script/dom/headers.rs:
```diff
@@ -217,6 +225,41 @@ impl Headers {
+
+    fn get_key_at_index(&self, n: u32) -> ByteString {
+        let sorted_header_vec = self.sort_header_list();
+        let key = sorted_header_vec[n as usize].1.clone();
```
I suspect this should be `.0` instead... do any test results change when that's corrected?

jeenalee replied:
Oh strange! Yes you are correct. I had it as 0, but I don't know what happened there... I'll fix that right now.

jdm commented:
There are a bunch of unrelated changes in the commit that implements Iterable; it would be best to remove those (at least into separate commits that can be removed from this PR before merging).

jdm commented:
#12819 should be merging within the next couple hours, so feel free to put this all until you can rebase your changes against master.

jeenalee commented:
Ok. I must have made mistakes with git. It wasn't my intention to group all the changes in the headers with the iterable implementation commit. I'll wait until the #12819 is merged into master.

jeenalee committed:
Implement iterable for Headers.

This commit implements iterable in DOM Headers based on iterable
implementation from #12819. Expected wpt results are updated as well.

jeenalee commented:
@jdm Rebased!

jdm commented:
@bors-servo: r+
Nice work!

bors-servo commented:
:pushpin: Commit e3624ed has been approved by `jdm`

<!-- @bors-servo r=jdm e3624edbf6329d63ed2f3716250ab4e8fc254f63 -->

bors-servo commented:
:hourglass: Testing commit e3624ed with merge 5b46a59...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/2625), [arm64](http://build.servo.org/builders/arm64/builds/2640), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2691), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2853), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2709), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2697), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2795), [windows-dev](http://build.servo.org/builders/windows-dev/builds/587)
